src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -0,0 +1,322 @@
+                final List<GATKRead> frags = Ordering.natural().reverse().onResultOf((GATKRead read) -> score(read)).immutableSortedCopy(byPairing.get(false));
+                if (!frags.isEmpty()) {
+                    reads.add(frags.get(0));                        //highest score - just emit
+                    for (final GATKRead record : Iterables.skip(frags, 1)) {  //lower   scores - mark as dups and emit
+                        record.setIsDuplicate(true);
+                        reads.add(record);
+                    }
+                }
+            } else {
+                // There are paired ends so we mark all fragments as duplicates.
+                for (final GATKRead record : byPairing.get(false)) {
+                    record.setIsDuplicate(true);
+                    reads.add(record);
+                }
+            }
+            return reads;
+        });
+    }
+
+
+    /**
+     * How to assign a score to the read in MarkDuplicates (so that we pick the best one to be the non-duplicate).
+     */
+    //Note: copied from htsjdk.samtools.DuplicateScoringStrategy
+    static int score(final GATKRead record) {
+        if (record == null) {
+            return 0;
+        } else {
+            int sum = 0;
+            for ( byte b : record.getBaseQualities() ) {
+                int i = (int)b;
+                if ( i >= MIN_BASE_QUAL ) {
+                    sum += i;
+                }
+            }
+            return sum;
+        }
+    }
+
+    /**
+     * (1) keyReadsByName: label each read with its read group and read name.
+     * (2) GroupByKey: group together reads with the same group and name.
+     * (3) keyPairedEndsWithAlignmentInfo:
+     *   (a) Sort each group of reads (see GATKOrder below).
+     *   (b) Pair consecutive reads into PairedEnds. In most cases there will only be two reads
+     *       with the same name. TODO: explain why there might be more.
+     *   (c) Label each read with alignment information: Library, reference index,
+     *       stranded unclipped start and reverse strand.
+     *   (d) Leftover reads are emitted, unmodified, as an unpaired end.
+     * (4) GroupByKey: Group PairedEnds that share alignment information. These pairs
+     *     are duplicates of each other.
+     * (5) markDuplicatePairs:
+     *   (a) For each group created by (4), sort the pairs by score and mark all but the
+     *       highest scoring as duplicates.
+     *   (b) Determine which duplicates are optical duplicates and increase the overall count.
+     */
+    static JavaRDD<GATKRead> transformReads(final SAMFileHeader header, final OpticalDuplicateFinder finder, final JavaRDD<GATKRead> pairs) {
+
+        JavaPairRDD<String, Iterable<GATKRead>> keyedReads =
+                pairs.filter(ReadUtils::readHasMappedMate).mapToPair(read -> new Tuple2<>(ReadsKey.keyForRead(header, read), read)).groupByKey();
```

Comment:
For the .filter, that should be added to the doc.
keyForRead needs it for the library for the read (also to get the reference index).